### ISSUE TITLE
get_params function is mandatory for provided ciphers

### ISSUE THREAD

**Issue by mattcaswell** [openssl/openssl#19110]
It appears that having a "get_params" function is mandatory for a provided cipher.

This is not checked for when we check that a cipher has all the minimum functions that it needs:

https://github.com/openssl/openssl/blob/3d623896eb50f5b15d3ef8f53b9f1e5c7546695a/crypto/evp/evp_enc.c#L1630-L1642

But the "fetch" still fails because it *must* be there when we cache the cipher constants (because otherwise `evp_cipher_cache_constants` fails:

https://github.com/openssl/openssl/blob/3d623896eb50f5b15d3ef8f53b9f1e5c7546695a/crypto/evp/evp_enc.c#L1647-L1651

This tripped up an author of a new provider here:

https://github.com/orgs/provider-corner/discussions/2

It seems to me that the way `evp_cipher_cache_constants` is written is that no particular constants need to be present, but "get_params" must be there (so a "get_params" function that supports no parameters would be sufficient). I am not convinced that this requirement to have "get_params" is deliberate. Probably it should be optional to have one (in which case no constants get cached). In fact the man page says that it is optional:

> A cipher algorithm implementation may not implement all of these functions. In order to be a consistent set of functions there must at least be a complete set of "encrypt" functions, or a complete set of "decrypt" functions, or a single "cipher" function. In all cases both the OSSL_FUNC_cipher_newctx and OSSL_FUNC_cipher_freectx functions must be present. All other functions are optional.

https://www.openssl.org/docs/man3.0/man7/provider-cipher.html

Either we should make "get_params" optional as per the documentation, or we should update the documentation to say that it is required.

**Comment by levitte** [comment]
> I am not convinced that this requirement to have "get_params" is deliberate.

It isn't, we've been sloppy checking if it's there or not.  The original idea is that it should be possible to do without, in which case there are no values to be cached.
However, come to think of it, this idea might not be feasible after all, since there are some values that we may need to have unconditionally, such as key size, block size, that sort of stuff.

> Either we should make "get_params" optional as per the documentation, or we should update the documentation to say that it is required.

I would prefer that it remains optional, but we need to check that it's feasible.

**Comment by t8m** [comment]
To me it does not make any sense to keep it as optional function. At least it should be possible to query the key size, block size, and iv size, if the algorithm has an iv.

**Comment by mattcaswell** [comment]
> To me it does not make any sense to keep it as optional function. At least it should be possible to query the key size, block size, and iv size, if the algorithm has an iv.

Unfortunately where we are now is that no parameter is listed as mandatory - they are all optional. Making a currently optional parameter mandatory would be a breaking change so its not something we can do in a minor release. So all parameters are optional, but the presence of the "get_params" function itself is mandatory. That doesn't make a lot of sense and I don't think it was what was originally intended. Making "get_params" optional is probably something we can do as a bug fix without breaking anything.

**Comment by levitte** [comment]
> To me it does not make any sense to keep it as optional function. At least it should be possible to query the key size, block size, and iv size, if the algorithm has an iv.

Well, depends.  If the application knows the cipher intimately, it would be unsurprising if it knows all those sizes, or whether an IV is needed or not.  In that case, it has zero need for those cached constants, it would just init, update, final and be done with it.
It is perfectly feasible for a provider to be documented with the necessary information and to expect the application author to read that and act accordingly.
Of course, it's *nice* if the provider can provide those parameters for you, but that doesn't mean it should be mandatory, however rough that may be.

**Comment by levitte** [comment]
I did a quick test with the `vigenere` toy, simply having its `get_params` do nothing...  It turns out badly, unfortunately...  among others, it seems that we depend on block size being given.

Intuitively, I'm not sure why we couldn't do without a given block size...  essentially, we would have to consider a cipher for which we don't know the block size a stream cipher, but that doesn't seem impossible for most all but ECB mode ciphers, or have I misunderstood that part?

**Comment by beldmit** [comment]
Stream ciphers have block size 1

**Comment by levitte** [comment]
> Stream ciphers have block size 1

Yes.  So our code could assume 1 if the cipher doesn't set a block size

**Comment by mattcaswell** [comment]
That would be a reasonable default, but if its wrong we will not pad correctly.

**Comment by t8m** [comment]
I really do not see the fuss of providing one more function returning a few constant params when writing a cipher provider. We could just document the de-facto state.

**Comment by levitte** [comment]
The original idea (at least in my mind) was that parameters handled with OSSL_PARAMS would always be optional.  That's a reason why it's documented as it is.  In particular, we might want to pay attention to this line:

https://github.com/openssl/openssl/blob/fc2be2d07acc0cfe954320c2491b8c5461cbef09/doc/man7/provider-cipher.pod?plain=1#L207

So, unless blocksize is relevant to some cipher, why should it care to give one?

**Comment by paulidale** [comment]
Didn't you earlier write the our code crashes if no block size is available?
That seems like a good reason to provide one.

**Comment by levitte** [comment]
> Didn't you earlier write the our code crashes if no block size is available? That seems like a good reason to provide one.

That can as well be seen as a bug in our code.

**Comment by levitte** [comment]
> That would be a reasonable default, but if its wrong we will not pad correctly.

I was debating with myself if that's something libcrypto *must* care about that...  but yeah, I guess it must at least say that there is buffer space for that, there doesn't seem to be a sane way out of that, at least at a quick code read.  If nothing else, this part is a glaring demonstration:

https://github.com/openssl/openssl/blob/fc2be2d07acc0cfe954320c2491b8c5461cbef09/crypto/evp/evp_enc.c#L710-L711

**Comment by paulidale** [comment]
> That can as well be seen as a bug in our code.

It is a bug in either our code or documentation.  It also guarantees that nobody is relying on documented behaviour.

**Comment by mattcaswell** [comment]
> block and key size are mandatory

Is key size mandatory? I didn't see anything to suggest that missing a key size would cause a crash at the moment?

**Comment by levitte** [comment]
> > block and key size are mandatory
> 
> Is key size mandatory?

Looks that way.

https://github.com/openssl/openssl/blob/fc2be2d07acc0cfe954320c2491b8c5461cbef09/crypto/evp/evp_enc.c#L233-L240
https://github.com/openssl/openssl/blob/fc2be2d07acc0cfe954320c2491b8c5461cbef09/crypto/evp/evp_enc.c#L248-L255

> I didn't see anything to suggest that missing a key size would cause a crash at the moment?

"crash" is a bit strong, but you get init failure or incorrect encryption because the key size that's passed to `->einit` / `->dinit` is invalid or smaller than intended (i.e. zero).

**Comment by mattcaswell** [comment]
> "crash" is a bit strong, but you get init failure or incorrect encryption because the key size that's passed to ->einit / ->dinit is invalid or smaller than intended (i.e. zero).

Hmmm. This is actually a bit of an issue because the thing that will fail here is the provider implementation. So, if a we suddenly  make key size a mandatory param that could actually result in a breaking change if the provider implementation ignores that key size that is passed back in the init call.

**Comment by levitte** [comment]
> > "crash" is a bit strong, but you get init failure or incorrect encryption because the key size that's passed to ->einit / ->dinit is invalid or smaller than intended (i.e. zero).
> 
> Hmmm. This is actually a bit of an issue because the thing that will fail here is the provider implementation. So, if a we suddenly make key size a mandatory param that could actually result in a breaking change if the provider implementation ignores that key size that is passed back in the init call.

Not really?  If their init functions ignored the key size before, they can continue to do so and continue to assume what they already assume about the key buffer.

**Comment by levitte** [comment]
> Not really? If their init functions ignored the key size before, they can continue to do so and continue to assume what they already assume about the key buffer.

Ah, I can actually test that

**Comment by mattcaswell** [comment]
> Not really? If their init functions ignored the key size before, they can continue to do so and continue to assume what they already assume about the key buffer.

My point is that if their init functions ignored the key size before, and their `get_param` function did not provide a key size param, then making key size a mandatory parameter is a breaking change for them.

**Comment by levitte** [comment]
Re the key size, not setting one in get_params, and ignoring in the provider's encryption/decryption init function works fine.

... But!

Our `openssl enc` makes assumptions based on `EVP_CIPHER_get_key_length()`:

https://github.com/openssl/openssl/blob/948c656c66a3846337a0262197766c80ec7c9e59/apps/enc.c#L494-L503
https://github.com/openssl/openssl/blob/948c656c66a3846337a0262197766c80ec7c9e59/apps/enc.c#L546-L553

**Comment by mattcaswell** [comment]
> Re the key size, not setting one in get_params, and ignoring in the provider's encryption/decryption init function works fine.

Right - so basic cipher functionality works. Other things that you might do with that cipher may or may not work, but there is at least some scenarios where making key len a mandatory parameter could cause existing applications to break.
